File: Java/LinkedList/Reverse_LinkedList.java
public class Reverse_LinkedList {
    private class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public ListNode reverseLinkedList(ListNode head) {

        ListNode curr = head;
        ListNode prev = null;

        while(curr != null) {
            ListNode nextTemp = curr.next;
            curr.next = prev;
            prev = curr;

            curr = nextTemp;
        }

        return prev;
    }
    // t: 0(N)
    // s: O(1)

    public ListNode reverseLinkedListRecursive(ListNode head) {

        if (head == null || head.next == null) {
            return null;
        }

        ListNode p = reverseLinkedListRecursive(head.next);
        head.next.next = head;
        head.next = null;
        return p;
    }

}

/**
 * Definition for singly-linked list.
 *
 */
//
// public class ListNode {
//      int val;
//      ListNode next;
//      ListNode() {}
//      ListNode(int val) { this.val = val; }
//      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
// }

File: Python/4_______LINKEDLIST_______/Reverse_LinkedList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # t: (n), s: (1)
    def reverseList(self, head: ListNode) -> ListNode:
        
        prev = None
        curr = head
        while curr:
            tmp = curr.next #remember next node
            curr.next = prev # REVERSE! None, first time around 
            prev = curr
            curr = tmp
        
        return prev